azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -526,3 +526,184 @@ def _handle_pipeline_event(self, event):
+    of the pipeline.
+
+    Unlike the TimeoutStage, this stage will never need to worry about cancelling
+    failed operations.  When an operation is retried at this stage, it is already
+    considered "failed", so no cancellation needs to be done.
+    """
+
+    def __init__(self):
+        super(RetryStage, self).__init__()
+        self.retry_intervals = {
+            pipeline_ops_mqtt.MQTTSubscribeOperation: 20,
+            pipeline_ops_mqtt.MQTTUnsubscribeOperation: 20,
+        }
+        self.waiting_to_retry = []
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _execute_op(self, op):
+        """
+        Send all ops down and intercept their return to "watch for retry"
+        """
+        self._send_op_down_and_intercept_return(
+            op=op, intercepted_return=self._on_intercepted_return
+        )
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _should_watch_for_retry(self, op):
+        """
+        Return True if this op needs to be watched for retry.  This can be
+        called before the op runs.
+        """
+        return op.__class__ in self.retry_intervals
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _should_retry(self, op, error):
+        """
+        Return True if this op needs to be retried.  This must be called after
+        the op completes.
+        """
+        if error:
+            if self._should_watch_for_retry(op):
+                if error.__class__ == transport_exceptions.PipelineTimeoutError:
+                    return True
+        return False
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _on_intercepted_return(self, op, error):
+        """
+        Handler which gets called when operations are complete.  This function
+        is where we check to see if a retry is necessary and set a "retry timer"
+        which can be used to send the op down again.
+        """
+        if self._should_retry(op, error):
+            self_weakref = weakref.ref(self)
+
+            @pipeline_thread.invoke_on_pipeline_thread_nowait
+            def do_retry():
+                this = self_weakref()
+                logger.info("{}({}): retrying".format(this.name, op.name))
+                del op.retry_timer
+                op.completed = False
+                this.waiting_to_retry.remove(op)
+                # Don't just send it down directly.  Instead, go through _execute_op so we get
+                # retry functionality this time too
+                this._execute_op(op)
+
+            interval = self.retry_intervals[op.__class__]
+            logger.warning(
+                "{}({}): Op needs retry with interval {} because of {}.  Setting timer.".format(
+                    self.name, op.name, interval, error
+                )
+            )
+
+            # if we don't keep track of this op, it might get collected.
+            self.waiting_to_retry.append(op)
+            op.retry_timer = Timer(self.retry_intervals[op.__class__], do_retry)
```

Comment:
>op.retry_timer [](start = 12, length = 14)

Dynamically applied attributes make me nervous. I'd like to see the base op class updated to set the `.retry_timer` to `None` in the `__init__()`